### PR TITLE
Fix async loading in Database constructor

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

- Move db initializations from the constructor to an async bootstrap 
- Because those properties are no longer guaranteed as defined, but they need to be defined to be a non-breaking change for typescript, I created get/set methods for everything that throw when used when not loaded. A simpler alternative would be non-null assertions, but it seemed better to provide predictability in throwing errors, instead of potentially unpredictable behavior. But anyone please comment with preference and reasons for why one may be better than the other.

### Why is it needed?

- We were calling async dialect.initialize methods synchronously so they weren't running before providing the connection as "ready"
- some other init operations such as createMigrationsProvider will need to become async to support typescript compilation or other steps
  
### How to test it?

Database functionality should remain unchanged.

A test has been added to ensure unexpected usage will throw an error.

### Related issue(s)/PR(s)

Based on #18010 by @DevDengChao
Co-authored-by: DevDengChao <2325690622@qq.com>

### CODE REVIEW

File: packages/core/database/src/index.ts [hunk 1]
```diff
@@ -30,30 +30,109 @@ export interface DatabaseConfig {
...
   config: DatabaseConfig;
 
   metadata: Metadata;
 
-  schema: SchemaProvider;
+  /**
+   * NOTE: The get/set methods are necessary to prevent a breaking change in typescript now that they are not set in the constructor
+   */
+
+  #connection?: Knex;
```
Comment (alexandrebodin) [inline packages/core/database/src/index.ts:41]: You can use the private keyword from TS that is much more readable 👍

File: packages/core/database/src/index.ts [hunk 2]
```diff
@@ -30,30 +30,109 @@ export interface DatabaseConfig {
...
+      throw new Error('Database.entityManager is not loaded');
+    }
+    return this.#entityManager;
+  }
+
+  set entityManager(value: EntityManager) {
+    this.#entityManager = value;
+  }
+
+  #bootstrapComplete = false;
```
Comment (alexandrebodin) [inline packages/core/database/src/index.ts:119]: should we use a "state": "init"|"ready"... ?

File: packages/core/database/src/migrations/index.ts [hunk 3]
```diff
@@ -70,11 +70,7 @@ const createUmzugProvider = (db: Database) => {
 
 // NOTE: when needed => add internal migrations for core & plugins. How do we overlap them with users migrations ?
 
-/**
- * Creates migrations provider
- * @type {import('.').createMigrationsProvider}
- */
-export const createMigrationsProvider = (db: Database): MigrationProvider => {
+export const createMigrationsProvider = async (db: Database): Promise<MigrationProvider> => {
```
Comment (alexandrebodin) [inline packages/core/database/src/migrations/index.ts:73]: Can you remind me why this should be async ?  FYI we will have a conflict in v5 here

Comment (innerdvations) [reply packages/core/database/src/migrations/index.ts:73]: this one needs to be async for [typescript migration compilation](https://github.com/strapi/strapi/pull/18010)

Comment (alexandrebodin) [reply packages/core/database/src/migrations/index.ts:73]: Can we delay this for v5 as I refactor this whole part and we might need to redo the change 🤔

Comment (innerdvations) [reply packages/core/database/src/migrations/index.ts:73]: Up to you, I'm happy to wait until v5 as long as we don't worry about not supporting typescript migrations in v4

Comment (alexandrebodin) [reply packages/core/database/src/migrations/index.ts:73]: Yeah I think that's fine delaying it.

**Pull request closed without merging**